### ISSUE TITLE
disable_dp hangs the training process

### ISSUE THREAD

**Issue by edifice1989** [microsoft/dp-transformers#56]
When I used the args --disable_dp True
the training seems not to be working anymore.It simply hangs without any loss logging and progress
with warning like /home/ec2-user/anaconda3/envs/dp-transformers-qlora/lib/python3.10/site-packages/prv_accountant/privacy_random_variables.py:188: RuntimeWarning: divide by zero encountered in scalar divide
  log_e0 = math.log(.5) + _log_erfc((i - z0) / (math.sqrt(2) * sigma))
/home/ec2-user/anaconda3/envs/dp-transformers-qlora/lib/python3.10/site-packages/prv_accountant/privacy_random_variables.py:189: RuntimeWarning: divide by zero encountered in scalar divide
  log_e1 = math.log(.5) + _log_erfc((z0 - j) / (math.sqrt(2) * sigma))
/home/ec2-user/anaconda3/envs/dp-transformers-qlora/lib/python3.10/site-packages/prv_accountant/privacy_random_variables.py:191: RuntimeWarning: invalid value encountered in scalar divide
  log_s0 = log_t0 + (i * i - i) / (2 * (sigma**2)) + log_e0
/home/ec2-user/anaconda3/envs/dp-transformers-qlora/lib/python3.10/site-packages/prv_accountant/privacy_random_variables.py:192: RuntimeWarning: divide by zero encountered in scalar divide
  log_s1 = log_t1 + (j * j - j) / (2 * (sigma**2)) + log_e1
/home/ec2-user/anaconda3/envs/dp-transformers-qlora/lib/python3.10/site-packages/prv_accountant/privacy_random_variables.py:191: RuntimeWarning: divide by zero encountered in scalar divide
  log_s0 = log_t0 + (i * i - i) / (2 * (sigma**2)) + log_e0
/home/ec2-user/anaconda3/envs/dp-transformers-qlora/lib/python3.10/site-packages/prv_accountant/privacy_random_variables.py:191: RuntimeWarning: invalid value encountered in scalar add
  log_s0 = log_t0 + (i * i - i) / (2 * (sigma**2)) + log_e0
/home/ec2-user/anaconda3/envs/dp-transformers-qlora/lib/python3.10/site-packages/prv_accountant/privacy_random_variables.py:192: RuntimeWarning: invalid value encountered in scalar add
  log_s1 = log_t1 + (j * j - j) / (2 * (sigma**2)) + log_e1
/home/ec2-user/anaconda3/envs/dp-transformers-qlora/lib/python3.10/site-packages/prv_accountant/privacy_random_variables.py:168: RuntimeWarning: invalid value encountered in scalar subtract
  return math.log1p(math.exp(a - b)) + b  # log1p(x) = log(x + 1)

Here is my command:
python research/fine_tune_llm_w_qlora/fine-tune-dp.py \
    --output_dir outputs \
    --model_name mistralai/Mistral-7B-v0.1 \
    --dataset_name sst2 \
    --sequence_len 128 \
    --per_device_train_batch_size 8 \
    --gradient_accumulation_steps 16 \
    --evaluation_strategy steps \
    --eval_steps 4 \
    --save_strategy no \
    --log_level info \
    --per_device_eval_batch_size 16 \
    --eval_accumulation_steps 1 \
    --seed 42 \
    --target_epsilon 8 \
    --target_delta 1e-5 \
    --per_sample_max_grad_norm 1.0 \
    --weight_decay 0.01 \
    --remove_unused_columns False \
    --num_train_epochs 3 \
    --logging_steps 4 \
    --max_grad_norm 0 \
    --lr_scheduler_type constant \
    --learning_rate 3e-4 \
    --disable_tqdm False \
    --dataloader_num_workers 2 \
    --lora_dim 4 \
    --lora_alpha 32 \
    --lora_dropout 0.0 \
    --enable_lora \
    --target_modules "['q_proj', 'v_proj']" \
    --label_names labels \
    --bf16 --disable_dp True

**Comment by huseyinatahaninan** [comment]
Yes, you are right. It seems that prv_accountant still thinks we are applying DP but with noise multiplier 0.0. This can be fixed but would you prefer switching to /fine-tune-nodp.py for training without DP? Or do you for some reason prefer staying in /fine-tune-dp.py but training with noise_multiplier 0.0 and per_sample_max_grad_norm infinity to simulate non-DP training?

**Comment by edifice1989** [comment]
I mainly want to isolate DP's influence on the quality of my training results. I got some weird nonsensical generationoutput  from dp-fined model. there is no error reported during training. I am scracthing my head. one quick question, does Opacus prefer to work with float16 or float32 dtype?

**Comment by huseyinatahaninan** [comment]
fp16 may be problematic but bf16 should be okay I think. if you'd like to continue with fine-tune-dp.py maybe you can try commenting out the parts that involve self.compute_prv_epsilon because it is setting noise to 0 and that's leading to error in prv epsilon calculation.

**Comment by edifice1989** [comment]
thank you, that is helpful. I will try that to see what happens: )

**Comment by edifice1989** [comment]
after masking all lines involved prv and set eps to 80, grad_norm like 30 to minic no dp model training.  it seems like the training process converged with loss steadily going down until flat. however, the generated output looks like the same as the base model, aka un-fine-tuned one. so my question is, can we directly use the fine-tuned model to do generation from research/fine_tune_llm_w_qlora/fine-tune-dp.py ? thx

**Comment by huseyinatahaninan** [comment]
you should be able to save the model and then later load it for generation, it seems to me that something may be wrong with how you load the model? Can you show you do perform the model loading and generation?

**Comment by edifice1989** [comment]
I replaced the HF classic trainer with SFTtrainer to do supervised fine-tuning. 
Below is the model serving part right after the model training on the fly. 
I disabled DP, generation works ok. but DP somehow caused a problem even with little noise injected.    
 # Add generation section after training
    logger.info("Starting text generation...")
    
    alpaca_prompt = """Below is an instruction that describes a task, paired with an input that provides further context. Write a response that appropriately completes the request.
    ### Instruction:
    {}
    ### Input:
    {}
    ### Response:
    {}"""


    CORPORATE_EMAIL_CATEGORIES = [
        "Project Updates",
        "Meeting Coordination",
        "HR Communications",
        "IT Support and Systems",
        "Financial Reports",
        "Team Collaboration",
        "Training and Development",
        "Policy Updates",
        "Client Communications",
        "Resource Requests",
        "Performance Reviews",
        "Company Announcements",
        "Department Updates",
        "Process Documentation",
        "Technical Documentation",
        "Sales and Marketing",
        "Research and Development",
        "Operations",
        "Employee Engagement"
    ]
    num_replicas_per_category = 1
    generated_examples = []

    # Set model to evaluation mode
    model.eval()

    # Generate examples for each category
    with torch.no_grad():
        for category in CORPORATE_EMAIL_CATEGORIES:
            for _ in range(num_replicas_per_category):
                prompt = alpaca_prompt.format(
                    f"write me some corporate  email example in the category of {category}",
                    "",
                    ""
                )
                
                inputs = tokenizer(prompt, return_tensors="pt", truncation=True, max_length=512)
                inputs = {k: v.to(model.device) for k, v in inputs.items()}
                
                outputs = model.generate(
                    **inputs,
                    max_new_tokens=128,
                    temperature=0.5,
                    top_p=0.95,
                    top_k=50,
                    pad_token_id=tokenizer.pad_token_id,
                    eos_token_id=tokenizer.eos_token_id,
                )
                
                generated_text = tokenizer.decode(outputs[0], skip_special_tokens=True)
                generated_examples.append({
                    "category": category,
                    "generated_text": generated_text
                })
                
                logger.info(f"Generated example for category: {category}")
                logger.info(f"Text: {generated_text}\n")

    # Save generated examples to a file
    output_file = os.path.join(train_args.output_dir, "generated_examples.txt")

**Comment by huseyinatahaninan** [comment]
ok I see the issue now. I think your training and generation are quite disjoint tasks so DP may be negatively affecting the way you are generating. From the command I see that you are training on `sst2` and this training will only learn to predict `positive` or `negative` given the text that `Sentence : <text> Label :`. You can try this and see that it should work. So this training is for classification, not for generation.

What you are doing in the generation part is to get some general instruction following behavior. This can also be done with DP but then you'd need to finetune on the right dataset with the corresponding format where you'd learn to predict the whole `### Response:` part. I hope this helps.
